### PR TITLE
fix(reference): stabilize LogSoftmax and SoftmaxCrossEntropyLoss

### PR DESCRIPTION
**Pull request by kadyrbekovhamit-cyber**
### Motivation and Context

The reference implementations currently materialize softmax probabilities and then take their logarithm. For finite float32 input `[[0, -104]]`, the suppressed probability underflows to zero, so `LogSoftmax` returns `[[0, -inf]]` and `SoftmaxCrossEntropyLoss` returns `[inf]` for target class 1.

The mathematically defined results are finite: `[[0, -104]]` and `[104]`. ONNX Runtime CPU returns those finite values for the same models. A stable backend can therefore appear to disagree with the reference exactly when it preserves the finite mathematical result.

### Change

Evaluate both operators directly in log space using max shifting:

```python
shifted = x - np.max(x, axis=axis, keepdims=True)
result = shifted - np.log(np.sum(np.exp(shifted), axis=axis, keepdims=True))
```

The expression is algebraically equivalent for finite inputs, preserves the input dtype, and retains the existing empty-input behavior.

### Tests

- Added a `LogSoftmax` regression with a finite 104-logit gap.
- Added a `SoftmaxCrossEntropyLoss` regression checking both the finite loss and optional `log_prob` output.
- Full `tests/python/reference_evaluator_test.py`: 444 passed, 4 skipped.

Classification: ordinary numerical correctness. This is not a security report and does not identify a defect in ONNX Runtime.

AI tool disclosure: OpenAI Codex assisted with reproducer drafting and patch preparation. The author independently executed the reproducer, reviewed the diff, and ran the reported test suite.

### CODE REVIEW

**Review by justinchuby — APPROVED** [review]

**Review by xadupre — APPROVED** [review]

File: onnx/reference/ops/op_softmax_cross_entropy_loss.py [hunk 1]
```diff
@@ -21,9 +21,10 @@ def softmaxcrossentropy(
 
     # compute log_softmax
     max_x = np.max(x, axis=1, keepdims=True)
-    exp_x = np.exp(x - max_x)
-    p = exp_x / np.sum(exp_x, axis=1, keepdims=True)
-    inp = np.log(p)
+    shifted_x = x - max_x
+    inp = shifted_x - np.log(
+        np.sum(np.exp(shifted_x), axis=1, keepdims=True)
+    )
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnx/reference/ops/op_softmax_cross_entropy_loss.py]: @kadyrbekovhamit-cyber could you address this comment? Thanks

[the remark this answers is not in the document]
Comment (kadyrbekovhamit-cyber) [reply onnx/reference/ops/op_softmax_cross_entropy_loss.py]: Addressed in 14270573: log_prob is now explicitly cast to the input dtype when it is created, so the weighted mean early-return path preserves T. I also added a float16 regression covering weight input, mean reduction, and the second log_prob output. The three focused stability and dtype tests pass, and lintrunner reports no issues. Thank you.
